pkg/operator/controller/certificate-publisher/controller.go

```diff
@@ -0,0 +1,160 @@
+	// secretIsInUse can look up whether a secret is referenced by any
+	// ingresscontrollers.
+	if err := operatorCache.IndexField(&operatorv1.IngressController{}, "defaultCertificateName", func(o runtime.Object) []string {
+		secret := controller.RouterEffectiveDefaultCertificateSecretName(o.(*operatorv1.IngressController), operandNamespace)
+		return []string{secret.Name}
+	}); err != nil {
+		return nil, fmt.Errorf("failed to create index for ingresscontroller: %v", err)
+	}
+
+	secretsInformer, err := operandCache.GetInformer(&corev1.Secret{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create informer for secrets: %v", err)
+	}
+	if err := c.Watch(&source.Informer{Informer: secretsInformer}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc:  func(e event.CreateEvent) bool { return reconciler.secretIsInUse(e.Meta) },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return reconciler.secretIsInUse(e.Meta) },
+		UpdateFunc:  func(e event.UpdateEvent) bool { return reconciler.secretIsInUse(e.MetaNew) },
+		GenericFunc: func(e event.GenericEvent) bool { return reconciler.secretIsInUse(e.Meta) },
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc:  func(e event.CreateEvent) bool { return reconciler.hasSecret(e.Meta, e.Object) },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return reconciler.hasSecret(e.Meta, e.Object) },
+		UpdateFunc:  func(e event.UpdateEvent) bool { return reconciler.secretChanged(e.ObjectOld, e.ObjectNew) },
+		GenericFunc: func(e event.GenericEvent) bool { return reconciler.hasSecret(e.Meta, e.Object) },
+	}); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// secretIsInUse returns true if the given secret is referenced by some
+// ingresscontroller.
+func (r *reconciler) secretIsInUse(meta metav1.Object) bool {
+	controllers := &operatorv1.IngressControllerList{}
+	lo := &client.ListOptions{}
+	lo.MatchingField("defaultCertificateName", meta.GetName())
+	if err := r.operatorCache.List(context.Background(), lo, controllers); err != nil {
+		log.Error(err, "failed to list ingresscontrollers", "related", meta.GetSelfLink())
+		return false
+	}
+	return len(controllers.Items) > 0
+}
+
+// hasSecret returns true if the effective default certificate secret for the
+// given ingresscontroller exists, false otherwise.
+func (r *reconciler) hasSecret(meta metav1.Object, o runtime.Object) bool {
+	ic := o.(*operatorv1.IngressController)
+	secretName := controller.RouterEffectiveDefaultCertificateSecretName(ic, r.operandNamespace)
+	secret := &corev1.Secret{}
+	if err := r.operandCache.Get(context.Background(), secretName, secret); err != nil {
+		if errors.IsNotFound(err) {
+			return false
+		}
+		log.Error(err, "failed to look up secret for ingresscontroller", "name", secretName, "related", meta.GetSelfLink())
```

Comment:
Error handling in predicates is tricky.  We can log a failure, but we cannot queue a retry from the predicate.  I hope that using the cache lets us avoid transient API failures.

Reply:
What's the risk of returning `true` if `operandCache.Get()` returns an unexpected error?

Reply:
The controller will not update `router-certs` till something else triggers a reconciliation.